File: Sample-Scripts/CSharp/List_all_aggregates_utilization_metrics.cs
//============================================================//
//        Sample code to list all the aggregates and          //
//    associated metrics, based on the space utilization      //
//============================================================//

using System;
using System.Net;
using System.IO;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using Newtonsoft.Json.Linq;  // A popular high-performance JSON framework for .NET, version 8.0.3 available in the lib directory

namespace VserverList
{
	class VserverList
	{
		static string HttpGet(string url, string uname, string pwd)
		{
			ServicePointManager.ServerCertificateValidationCallback = delegate(object s, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors){ return true; };

			HttpWebRequest req = WebRequest.Create(url) as HttpWebRequest;
			req.Credentials = new NetworkCredential(uname, pwd);
			string result = null;
			using (HttpWebResponse resp = req.GetResponse() as HttpWebResponse)
			{
				StreamReader reader = new StreamReader(resp.GetResponseStream());
				result = reader.ReadToEnd();
			}
			return result;
		}
		 public static void Main()
        {
            try
            {
				int i = 1;
				string cluster_get_url,metrics_get_url,cluster_response,metrics_response,cluster_ip=null,throughput=null,latency=null,total_transfers=null;
				float aggr_size_GB;
				
				Console.Write("Enter the IP address of the API Services server and the port number (x.x.x.x:port): ");
				string API_url = Console.ReadLine();
				Console.Write("Enter the API-S username: ");
				string API_username = Console.ReadLine();
				Console.Write("Enter the API-S password: ");
				string API_password = Console.ReadLine();
				
                string aggr_get_url = "https://" + API_url + "/api/1.0/ontap/aggregates?sortBy=size_used_percent&atomLinks=true"; // lists all the aggregates in ascending order of used size(%)			
				string aggr_response = HttpGet(aggr_get_url,API_username,API_password);
				dynamic aggregate = JObject.Parse(aggr_response);
				if(aggregate.result.total_records < 1)
				{  
					Console.WriteLine("No aggregates found");
				}
				else
				{
					Console.WriteLine("List of aggregates: " );
					Console.WriteLine("No. Name\t\tCluster IP\t\tSize used(%)\t\tSize Available(GB) Throughput\t\tLatency\t\tTotal Transfers");
					
					foreach(dynamic aggr_record in aggregate.result.records)
					{
						foreach (dynamic Link in aggr_record.links) // Use the atomic links to obtain details about the hosting cluster and aggregate metrics
						{
							if (Link.rel == "cluster") // Navigate to the cluster to obtain the IP address of the cluster
							{
								cluster_get_url = Link.href;
								cluster_response = HttpGet(cluster_get_url,API_username,API_password);
								dynamic aggr_cluster = JObject.Parse(cluster_response);
								cluster_ip = aggr_cluster.result.records[0].management_ip;
							}
							else if (Link.rel == "metrics") // Navigate to obtain the aggregate metrics
							{
								metrics_get_url = Link.href;
								metrics_response = HttpGet(metrics_get_url,API_username,API_password);
								dynamic aggr_metrics = JObject.Parse(metrics_response);
								foreach (dynamic metric in aggr_metrics.result.records[0].metrics)
								{
									if (metric.name == "throughput")
									{
										throughput = metric.samples[0].value + " " + metric.unit;
									}
									else if (metric.name == "latency")
									{
										latency = metric.samples[0].value + " " + metric.unit;
									}
									else if (metric.name == "total_transfers")
									{
										total_transfers = metric.samples[0].value + " " + metric.unit;
									}
								}
							}
						}
						aggr_size_GB = ((float)(aggr_record.size_avail))/1024/1024/1024;
						Console.WriteLine(i + " " + aggr_record.name + "\t\t"+ cluster_ip +"\t\t" + aggr_record.size_used_percent + "\t\t" + aggr_size_GB + "\t\t" + throughput + "\t\t" + latency + "\t\t" + total_transfers);
						i++;
					}
				}				
            }
            catch (Exception e) {
                System.Console.WriteLine(e.Message);
            }
        }
    }
}

File: Sample-Scripts/Python/List_all_aggregates_utilization_metrics.py
##########################################################
##        This program lists all the aggregates         ##
##        and associated metrics, based on the         	##
##                  space utilization                   ##
##########################################################
import requests
import json
import warnings

warnings.filterwarnings("ignore")
API_address = raw_input("Enter the IP address of the API Services server and the port number (x.x.x.x:port): ")
API_username = raw_input("Enter the API-S username: ")
API_password = raw_input("Enter the API-S password: ")

aggr_url= "https://" + API_address + "/api/1.0/ontap/aggregates?sortBy=size_used_percent&atomLinks=true" # lists all the aggregates in ascending order of used size(%)
APIAggrResponse = requests.get(aggr_url,auth=(API_username,API_password), verify=False)

if(APIAggrResponse.ok):
		print "----------------------------------------------------------------------\n\t\t\tList of aggregates\n----------------------------------------------------------------------"
		print "No.\tName\t\tCluster IP\t\tSize used(%)\tSize available(GB) Throughput\t\tLatency\t\tTotal Transfers"
		APIAggrData = json.loads(APIAggrResponse.content) # Loads (Load String) takes a Json file and converts into python data structure (dict or list, depending on JSON)
		i=1
		for AggrData in APIAggrData["result"]["records"]:
				for Link in AggrData["links"]: # Use the atomic links to obtain details about the hosting cluster and aggregate metrics
						if str(Link["rel"]) == "cluster": 
								cluster_url = Link["href"] # Navigate to the cluster to obtain the IP address of the cluster
								APIClusterResponse = requests.get(cluster_url,auth=(API_username,API_password), verify=False)
								if(APIClusterResponse.ok):
										APIClusterData = json.loads(APIClusterResponse.content)
										for ClusterData in APIClusterData["result"]["records"]:
												cluster_ip = str(ClusterData["management_ip"])
												
								else:
										APIClusterResponse.raise_for_status()
										
						if str(Link["rel"]) == "metrics": 
								metrics_url = Link["href"] # Navigate to obtain the aggregate metrics
								APIMetricsResponse = requests.get(metrics_url,auth=(API_username,API_password), verify=False)
								if(APIMetricsResponse.ok):
										APIMetricsData = json.loads(APIMetricsResponse.content)
										for MetricsData in APIMetricsData["result"]["records"]:
												for Metrics in MetricsData["metrics"]:
														if Metrics["name"] == "throughput":
															metrics_throughput = str(Metrics["samples"][0]["value"]) + " " + str(Metrics["unit"])
														if Metrics["name"] == "latency":
															metrics_latency = str(Metrics["samples"][0]["value"]) + " " + str(Metrics["unit"])
														if Metrics["name"] == "total_transfers":
															metrics_transfers = str(Metrics["samples"][0]["value"]) + " " + str(Metrics["unit"])
								else:										
										APIMetricsResponse.raise_for_status()

				print "%d\t%s\t\t%s\t\t%s\t\t%.2f\t\t%s\t%s\t%s" %(i,str(AggrData["name"]),cluster_ip,str(AggrData["size_used_percent"]),float(AggrData["size_avail"])/1024/1024/1024,metrics_throughput,metrics_latency,metrics_transfers)
				i = i+1

else:        
        APIAggrResponse.raise_for_status()
